src/main/java/com/google/sps/data/Room.java

```diff
@@ -14,19 +14,104 @@
 import java.util.stream.Collectors;
 import java.util.Map;
 import java.util.stream.*;
+import com.google.appengine.api.datastore.KeyFactory;
 
+//Object repsresenting the Room that the chat and video streaming will be in
+//TODO: Restructure so logic is more split up
 public class Room {
+    private static final int MAX_MESSAGES = 10;
     private static final String ROOM_ENTITY = "Room";
     private static final String MEMBERS_PROPERTY = "members";
     private static final String MESSAGES_PROPERTY = "messages";
     private static final String VIDEOS_PROPERTY = "videos";
     private static DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     private List<Member> members;
-    private LinkedList<Message> messages;
+    private Queue<Message> messages;
     private Queue<Video> videos;
+    private Key roomKey;
+
+    //Room factory function
+    public static Room createRoom(List<Member> members, Queue<Video> videos, Queue<Message> messages){
+        return new Room(members, videos, messages, null);
+    }
+
+    //Creates a room object from a Datastore Key
+    public static Room fromRoomId(long roomId) {
+        Key roomKey = KeyFactory.createKey("Room", roomId);
+        try {
+            return Room.fromRoomKey(roomKey));
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.toString());
+        }
+        return null;
+    }
+
+    //Turns a Room key into a Room object
+    private static Room fromRoomKey(Key roomKey) {
+        Entity roomEntity = datastore.get(roomKey);
+        Map<String, Object> properties = roomEntity.getProperties();
+        List<Member> memberList = 
+        ((ArrayList<EmbeddedEntity>) properties.get(MEMBERS_PROPERTY)).stream().map(Member::fromEmbeddedEntity).collect(Collectors.toCollection(ArrayList::new));
+        Queue<Video> videoQueue = 
+        ((ArrayList<EmbeddedEntity>) properties.get(VIDEOS_PROPERTY)).stream().map(Video::fromEmbeddedEntity).collect(Collectors.toCollection(LinkedList::new));
+        Queue<Message> messageQueue = (Queue<Message>) properties.get(MESSAGES_PROPERTY) != null ?
+        ((ArrayList<EmbeddedEntity>) properties.get(MESSAGES_PROPERTY)).stream().map(Message::fromEmbeddedEntity).collect(Collectors.toCollection(LinkedList::new)) : new LinkedList();
+        return new Room(memberList, videoQueue, messageQueue, roomKey);
+    }
+
+    //Turns the Room object into a datastore entity
+    private static Entity toEntity(Room room){
+        Entity newRoom = room.roomKey == null ? new Entity(ROOM_ENTITY) : new Entity(ROOM_ENTITY, room.roomKey);
+        newRoom.setProperty(MEMBERS_PROPERTY, room.getMembersAsEmbeddedEntities());
+        newRoom.setProperty(VIDEOS_PROPERTY, room.getVideosAsEntities());
+        newRoom.setProperty(MESSAGES_PROPERTY, room.getMessagesAsEntities());
+        return newRoom;
+    }
+
+    //Takes a room object and puts it into datastore
+    public Key toDatastore(){
```

Comment:
We shouldn't expose datastore related classes above. Let's change this to Key.getId()

Reply:
If you mean returning the id which is a long, I have already done that in the auth backend branch